polynote-server/src/main/scala/polynote/server/auth/IdentityProvider.scala

```diff
@@ -0,0 +1,166 @@
+  implicit val encoder: Encoder[PermissionType] = Encoder.encodeString.contramap(_.encoded)
+
+  private val setStringDecoder: Decoder[Set[PermissionType]] = Decoder.decodeString.flatMap {
+    case "all" => Decoder.const(All)
+    case str   => fromString(str).fold(Decoder.failedWithMessage, Decoder.const).map(Set(_))
+  }
+
+  implicit val setDecoder: Decoder[Set[PermissionType]] = setStringDecoder or Decoder.decodeSet(decoder)
+}
+
+sealed abstract class Permission(val permissionType: PermissionType)
+
+object Permission {
+  case class ReadNotebook(path: String) extends Permission(PermissionType.ReadNotebook)
+  case class ModifyNotebook(path: String) extends Permission(PermissionType.ModifyNotebook)
+  case class ExecuteCell(path: String, id: CellID) extends Permission(PermissionType.ExecuteCell)
+  case class CreateNotebook(path: String) extends Permission(PermissionType.CreateNotebook)
+  case class DeleteNotebook(path: String) extends Permission(PermissionType.DeleteNotebook)
+
+  case class PermissionDenied(permission: Permission, reason: String) extends Throwable(s"Permission denied: $permission ($reason)")
+}
+
+trait IdentityProvider {
+  def identityProvider: IdentityProvider.Service
+}
+
+object IdentityProvider {
+
+  trait Service {
+    /**
+      * Any special endpoints or pages that the provider has to mount, if there are any.
+      */
+    def authRoutes: Option[Routes]
+
+    /**
+      * Check the authorization for a request, failing with a Response (if not authorized) or succeeding with an
+      * Identity (if authorized) or None to indicate authorized anonymous access
+      */
+    def checkAuth(req: Request[Task]): ZIO[BaseEnv with Config, Response[Task], Option[Identity]]
+
+    /**
+      * Check whether the given identity (or None for anonymous) has the given permission. If not, the returned task
+      * should fail with PermissionDenied.
+      */
+    def checkPermission(ident: Option[Identity], permission: Permission): ZIO[BaseEnv with Config, Permission.PermissionDenied, Unit]
+  }
+
+  val noneService: Service = new Service {
+    def authRoutes: Option[Routes] = None
+    def checkAuth(req: Request[Task]): ZIO[BaseEnv with Config, Response[Task], Option[Identity]] = ZIO.succeed(None)
+    def checkPermission(ident: Option[Identity], permission: Permission): ZIO[BaseEnv with Config, Permission.PermissionDenied, Unit] = ZIO.unit
+  }
+
+  val none: IdentityProvider = of(noneService)
+
+  def of(service: Service): IdentityProvider = new IdentityProvider {
+    def identityProvider: Service = service
+  }
+
+  private lazy val loaders: Map[String, ProviderLoader] = {
+    import scala.collection.JavaConverters._
+    val loaders = ServiceLoader.load(classOf[ProviderLoader]).iterator.asScala.toList
+    loaders.map(l => l.providerKey -> l).toMap
+  }
+
+  def find(config: AuthProvider): RIO[BaseEnv with Config, IdentityProvider.Service] = ZIO(loaders)
```

Comment:
Per our discussion yesterday, should this be `effectBlocking(loaders)`?

Reply:
Yes, you're right.